### PR TITLE
[ONNX] Support registering custom export for prim::PythonOp from torch.autograd.Function (#55630)

### PR DESCRIPTION
**Pull request by BowenBao**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57830 [ONNX] Support index_add_ function. (#56867)
* #57829 [ONNX] Handling incorrect format for example_outputs  (#55802)
* #57605 [ONNX] Fix split export in opset13 (#56277)
* #57604 [ONNX] Handle mixed mask, index input for index_put
* #57603 [ONNX] ListUnpack on dynamic tensor list (#56592)
* #57602 [ONNX] Handle NoneType in Assign Output Shapes (#54623)
* #57601 [ONNX] Process const folding progressively when converts to ONNX (#54569)
* **#57600 [ONNX] Support registering custom export for prim::PythonOp from torch.autograd.Function (#55630)**
* #57599 Add support to to(device) op. (#56857)
* #57598 onnx.symbolic_helper.parse_args: document and clean up (#56956)
* #57597 [ONNX] Special post process for onnx::Cast and onnx::ConstantOfShape shape type inference (#55962)
* #57596 Add a new operator for fill_() function. (#56859)
* #57595 [ONNX] Warning when using __len__ to calculate tensor shape (#55151)
* #57594 [ONNX] Support .item() export & NumberType to tensor conversion (#55697)

Demo script:

```python
import torch

class MyReLU(torch.autograd.Function):
    @staticmethod
    def forward(ctx, input, scalar_tuple, scalar, scalar_list):
        ctx.save_for_backward(input)
        return input.clamp(min=scalar)
    @staticmethod
    def backward(ctx, grad_output):
        input, = ctx.saved_tensors
        grad_input = grad_output.clone()
        grad_input[input < 0] = 0
        return grad_input

class MyModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.linear_a = torch.nn.Linear(2, 2)
        self.linear_b = torch.nn.Linear(2, 2)
        self.relu = MyReLU.apply
    def forward(self, x):
        h = self.linear_a(x)
        h = self.relu(h, (5, 3), 2, [1, 2, 3])
        h = self.linear_b(h)
        return h

"""
User define how to export prim::PythonOp into custom op.
"""
def symbolic_pythonop(g, n, *args, **kwargs):
    # Print information:
    print('arguments of ', kwargs['name'], ':')
    print('original node: ', n)
    for i, out in enumerate(n.outputs()):
        print('original output {}: {}, requires grad: {}'.format(i, out, out.requiresGrad()))
    import torch.onnx.symbolic_helper as sym_helper
    for i, arg in enumerate(args):
        print('arg {}: {}, requires grad: {}'.format(i, arg, arg.requiresGrad() if sym_helper._is_value(arg) else False))
    for k, v in kwargs.items():
        print('key: ', k, ' v: ', v)

    # TODO: all inputs (tensors and scalars) are in args.
    #       backend can define CustomDomain::PythonOp and how info are stored however it deem fit.
    return g.op("CustomDomain::PythonOp", args[0], name_s=kwargs['name'])

torch.onnx.register_custom_op_symbolic("::prim_PythonOp", symbolic_pythonop, 9)

# Define input.
x = torch.tensor([[0.3971, 0.7544],
                  [0.5695, 0.4388]], requires_grad=True)

model = MyModule()
# Forward.
y = model(x)

torch.onnx.export(model, (x,), 'model.onnx', opset_version=12, verbose=True)
```

Co-authored-by: BowenBao <bowbao@microsoft.com>

Differential Revision: [D28393528](https://our.internmc.facebook.com/intern/diff/D28393528)

### CODE REVIEW

**Comment by SplitInfinity** [comment]
@SplitInfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28393528).
